Console

revnode opened the pull request:
Milestone 2 work.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

novabyte commented on console/ui/src/store/storage/sagas.ts:
```diff
@@ -0,0 +1,295 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Same as above.

mofirouz commented:
In addition to what @novabyte mentioned, I've found the following when running the dashboard via `yarn serve` and the server on port 80, using Safari and Chrome 73.

Notice - You'll need to run the server with the additional flag `--console.max_message_size_bytes 50000000` to allow for large values to be sent to the serve. We'll adjust defaults.

1. When logging in, see the message in Chrome 73: "Unchecked runtime.lastError: The message port closed before a response was received."

2. When updating a storage record fails, dashboard tries to load Storage view with a bad URL format. "Fetch failed loading: GET "http://127.0.0.1/v2/console/storage?"" and it fails, thus kicked to the login screen.

3. Stop refreshing blindly to "/login" or errors. Only kick to login on error code 401, otherwise please leave the page as it is. I think this goes hand in hand with issue 5 below.

4. Timestamp of updated Storage objects do not refresh when the update is done for the first time.

5. The login error is wiped due to a page reload. See attached GIF.
![badusername](https://user-images.githubusercontent.com/137265/55687447-f4318b00-5964-11e9-9432-facde5346bd5.gif)

mofirouz commented:
6. Bad layout on Configuration page. See image below.
7. Missing values on the configuration page (although they are available in the exported YAML file).
<img width="1150" alt="missingconfig" src="https://user-images.githubusercontent.com/137265/55687467-31961880-5965-11e9-9cae-a11a787a9f77.png">

8. Bad layout of warnings. See image (although may be fixed with what @novabyte said above). I think it is best to combine all warnings into one giant yellow box?
<img width="1378" alt="toomuchyellow" src="https://user-images.githubusercontent.com/137265/55687482-512d4100-5965-11e9-9fd8-d8fa7d603d42.png">

mofirouz commented:
9. When filtering storage by user ID, and then updating a given object, the view is punted back to the storage listing view and have forgotten the filtered user ID. Super annoying.

revnode commented on console/ui/src/routes/configuration/index.tsx:
```diff
@@ -41,7 +96,12 @@ class Configuration extends Component
               <Field.Body>
                 <Control>
-                  <Input static type="text" placeholder="(empty)" value="nakama-0" />
+                  <Input
```
Yup, done.

revnode commented on console/ui/src/routes/configuration/index.tsx:
```diff
@@ -233,15 +378,45 @@ class Configuration extends Component
               </Field.Body>
             </Field>
+            <br /><br />
+            {
```
Yup, done.

revnode commented on console/ui/src/store/configuration/sagas.ts:
```diff
@@ -0,0 +1,51 @@
+{
+  try
+  {
+    const nakama = NakamaApi({
```
Yup, done.

revnode commented:
> When logging in, see the message in Chrome 73: "Unchecked runtime.lastError: The message port closed before a response was received."

Not seeing this error on Chrome. Maybe it got resolved?

> When updating a storage record fails, dashboard tries to load Storage view with a bad URL format. "Fetch failed loading: GET "http://127.0.0.1/v2/console/storage?"" and it fails, thus kicked to the login screen.
> Stop refreshing blindly to "/login" or errors. Only kick to login on error code 401, otherwise please leave the page as it is. I think this goes hand in hand with issue 5 below.

Yup, fixed.

> Timestamp of updated Storage objects do not refresh when the update is done for the first time.

I can't check this because I'm getting the `{"error":"Requires a valid JSON object value.","message":"Requires a valid JSON object value.","code":3}` error from the backend when I try to update.

> The login error is wiped due to a page reload. See attached GIF.
> Bad layout on Configuration page. See image below.
> Missing values on the configuration page (although they are available in the exported YAML file).
> Bad layout of warnings. See image (although may be fixed with what @novabyte said above). I think it is best to combine all warnings into one giant yellow box?

Fixed.

> When filtering storage by user ID, and then updating a given object, the view is punted back to the storage listing view and have forgotten the filtered user ID. Super annoying. I think this may be fixed with better usage of History API rather than href.location.

Yup, fixed.

mofirouz commented:
I can't test your final commit because I get this error when running `yarn install && yarn build` or `yarn serve`:

```
/Users/mo/Documents/go/src/github.com/heroiclabs/nakama/console/ui/src/routes/storage/index.tsx
TypeScript error: Type 'string | string[]' is not assignable to type 'string'.
  Type 'string[]' is not assignable to type 'string'.  TS2322

    73 |     if(query.user_id)
    74 |     {
  > 75 |       (document.getElementById('user_id') as HTMLInputElement).value = query.user_id;
       |       ^
    76 |     }
    77 |     this.props.fetchManyRequest(query);
    78 |   }
```

mofirouz commented:
Also what is the value you are trying to save when getting: 

```{"error":"Requires a valid JSON object value.","message":"Requires a valid JSON object value.","code":3}
```

Because yesterday I tried to update a pre-existing value and it worked. Are you trying to save an empty value field because it should always be saved as `{}`.

revnode commented:
@mofirouz Typescript doesn't always show the errors while running it, fixed. As far as what I'm pushing for `value`, I just change an integer from like a 1 to a 2, nothing major, and it spits out that error every time.

mofirouz commented:
@revnode I think we need to make a small tweak to the way CSV exports are done to make it follow RFC4180. Instead of escaping double quotes like this `\"`, RFC4180 says that double quotes should be escaped with another double quote (i.e. `""`). 

Let me know if this is difficult to do and I'll do some magic on the server to handle this, but I prefer we stick to RFC4180. Also, I don't think you need to put each field inside a `"`.

mofirouz commented:
To quote the godocs:

```
Fields which start and stop with the quote character " are called
quoted-fields. The beginning and ending quote are not part of the
field.

The source:

 normal string,"quoted-field"

results in the fields

 {`normal string`, `quoted-field`}

Within a quoted-field a quote character followed by a second quote
character is considered a single quote.

 "the ""word"" is true","a ""quoted-field"""

results in

 {`the "word" is true`, `a "quoted-field"`}

Newlines and commas may be included in a quoted-field

 "Multi-line
 field","comma is ,"

results in

 {`Multi-line
 field`, `comma is ,`}
```
